### PR TITLE
Support build types

### PR DESCRIPTION
**Pull request by michaem**
Hey guys, here is a new version API of supporting build types based on :binary target type and self semantic.
For, example, we will have
```
:binary-debug,
:binary-release,
:binary-custom
```
And for each module we will have a simple configuration, as
```
buildConfiguration = BuildConfiguration(
        name = DebugBuildName(),
        singinConfig = DebugSigningConfig(
            keystoreFile = KeystoreFile("debug-keystore.properties")
        )
    )
```
Also I wanna to incapsulate some boilerplate of setting build types, configs names into type-safe classes.

Need your opinion, before we'll move to deep develop)

@KarenkovID You told about placing configs value into res directory of binary module. Could you plz give more details? How it will be looks for you? In my mind, I prefer to hold configurable build values more close to script file, but may be I mistake)

### CODE REVIEW

File: application/binary/build.gradle.kts [hunk 1]
```diff
@@ -23,5 +25,44 @@ androidBinary(
...
+    /**
+     * For :binary-custom
+     */
+    buildConfiguration = BuildConfiguration(
+        name = BuildName(
+            name = "custom",
+            appIdSuffix = ".custom"
+        ),
+        singinConfig = SigningConfig(
+            name = "Custom",
```
Comment (stepango) [inline application/binary/build.gradle.kts:64]: Why we heed a name here? I don't think we want to reuse is somwhere

File: application/binary/build.gradle.kts [hunk 2]
```diff
@@ -23,5 +25,44 @@ androidBinary(
...
+        )
+    ),
+
+    /**
+     * For :binary-custom
+     */
+    buildConfiguration = BuildConfiguration(
+        name = BuildName(
+            name = "custom",
+            appIdSuffix = ".custom"
```
Comment (stepango) [inline application/binary/build.gradle.kts:61]: we can use `packageName` instead

Comment (michaem) [reply application/binary/build.gradle.kts:61]: Good point!

File: application/binary/build.gradle.kts (hunk 1 above)
Comment (michaem) [reply application/binary/build.gradle.kts:64]: Yea, I checked. We don't need to pass additional name for configs, especially for different binaries.

**Pull request closed without merging**